File: 084. Largest Rectangle in Histogram.scala
/*
Given n non-negative integers representing the histogram's bar height where the width of each bar is 1, find the area of largest rectangle in the histogram.


Above is a histogram where width of each bar is 1, given height = [2,1,5,6,2,3].


The largest rectangle is shown in the shaded area, which has area = 10 unit.

For example,
Given heights = [2,1,5,6,2,3],
return 10.
*/

def largestRectangleArea(heights: Array[Int]): Int = {
    val stack = new scala.collection.mutable.Stack[Int]();
    var max = 0;
    for(i<- 0 to heights.length){
        var current = if(i<heights.length) heights(i) else 0;
        while(!stack.isEmpty && heights(stack.top)>current){
            var height = heights(stack.top)
            stack.pop();
            var length = if(stack.isEmpty) (i) else (i-1-stack.top)
            max = math.max(height*length,max);
        }
        stack.push(i);
    }        
    return max;
}

largestRectangleArea(Array(2,1,5,6,2,3))

File: 084. Largest Rectangle in Histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        if heights is None or len(heights)==0:
            return 0
        res = 0
        stack = []
        heights.append(0)
        for i in range(0,len(heights)):
            res = max(res,self.push(stack,i,heights))
        del heights[-1]
        return res
    
    def push(self,stack,i,nums):
        x = nums[i]
        res =0
        if len(stack)==0 or x>= nums[stack[-1]]:
            stack.append(i)
            return res
        else:
            while len(stack)>0 and nums[stack[-1]] > x:
                j = stack.pop()
                weigth = i - stack[-1] -1 if len(stack)>0 else i
                res = max(res,weigth*nums[j])
            stack.append(i)
            return res
